File: Easy/0035_search-insert-position/search-insert-position.py
#!/bin/python3

from typing import List, Tuple

class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        left : int = 0
        right : int = len(nums) - 1

        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] == target:
                return mid
            elif (nums[mid] > target):
                right = mid - 1
            else:
                left = mid + 1
        return left

def main():
    tests : List[Tuple[List[int], int]] = [
        ([1,3,5,6], 5),
        ([1,3,5,6], 2),
        ([1,3,5,6], 7),
    ]

    for arr, target in tests:
        print("---\nnums= ", arr, "\ntarget= ", target, "\nres= ", Solution().searchInsert(arr, target))

if __name__ == "__main__":
    main()

File: Easy/0035_search-insert-position/search-insert-position.go
package main

import (
	"fmt"
)

func searchInsert(nums []int, target int) int {
	left := 0
	right := len(nums) - 1

	for left <= right {
		mid := left + (right - left) / 2
		if nums[mid] == target {
			return mid
		} else if nums[mid] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return left
}

type testCase struct {
	arr []int
	target int
}

func main() {
	tests := []testCase{
		{arr: []int{1,3,5,6}, target:5},
		{arr: []int{1,3,5,6}, target:2},
		{arr: []int{1,3,5,6}, target:7},
	}

	for _,tc := range tests {
		fmt.Printf("---\nnums= %v\ntarget= %d\nres= %d\n", tc.arr, tc.target, searchInsert(tc.arr, tc.target))
	}
}
